oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java

```diff
@@ -81,47 +86,50 @@ public void setAllowFormEncodedBodyParameter(boolean allowFormEncodedBodyParamet
 	 *
 	 * @param allowUriQueryParameter if the URI query parameter is supported
 	 */
-	public void setAllowUriQueryParameter(boolean allowUriQueryParameter) {
+	public void setAllowUriQueryParameter(final boolean allowUriQueryParameter) {
 		this.allowUriQueryParameter = allowUriQueryParameter;
 	}
 
-	private static String resolveFromAuthorizationHeader(HttpServletRequest request) {
-		String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
-		if (StringUtils.hasText(authorization) && authorization.startsWith("Bearer")) {
-			Matcher matcher = authorizationPattern.matcher(authorization);
-
+	protected String resolveFromAuthorizationHeader(final HttpServletRequest request) {
+		final String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
+		if (authorization != null && authorization.startsWith("Bearer")) {
+			final Matcher matcher = AUTHORIZATION_PATTERN.matcher(authorization);
 			if (!matcher.matches()) {
-				BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INVALID_TOKEN,
+				final BearerTokenError error = new BearerTokenError(BearerTokenErrorCodes.INVALID_TOKEN,
 						HttpStatus.UNAUTHORIZED,
 						"Bearer token is malformed",
 						"https://tools.ietf.org/html/rfc6750#section-3.1");
 				throw new OAuth2AuthenticationException(error);
 			}
-
 			return matcher.group("token");
 		}
 		return null;
 	}
 
-	private static String resolveFromRequestParameters(HttpServletRequest request) {
-		String[] values = request.getParameterValues("access_token");
-		if (values == null || values.length == 0)  {
+	protected String resolveFromRequestParameters(final HttpServletRequest request) {
+		// By this point there is either zero or one value
+		final String[] values = request.getParameterValues(ACCESS_TOKEN_PARAM_NAME);
+		if (values == null || values.length == 0 || !isParameterTokenSupportedForRequest(request)) {
 			return null;
 		}
+		return values[0];
+	}
 
-		if (values.length == 1) {
-			return values[0];
+	protected boolean isParameterTokenSupportedForRequest(final HttpServletRequest request) {
+		if (isTokenInUriQueryParameter(request)) {
+			return this.allowUriQueryParameter;
+		} else {
+			return this.allowFormEncodedBodyParameter
+					&& MediaType.APPLICATION_FORM_URLENCODED.equals(MediaType.valueOf(request.getContentType()))
+					&& !RequestMethod.GET.equals(RequestMethod.valueOf(request.getMethod()));
```

Comment:
This part is debatable because as per spec [2.2.  Form-Encoded Body Parameter](https://tools.ietf.org/html/rfc6750#section-2.2) "`The client MUST NOT use this method unless
   all of the following conditions are met:`", it does not mention it must be considered as `invalid_request`